src/ndarray/ndarray.cc

```diff
@@ -1582,8 +1582,14 @@ static const uint32_t NDARRAY_V2_MAGIC = 0xF993fac9;
 
 void NDArray::Save(dmlc::Stream *strm) const {
   // TODO(junwu): Support this after NumPy operators are merged
-  CHECK(!Imperative::Get()->is_np_shape())
-      << "Saving ndarray within the scope of np_shape is not supported.";
+  if (Imperative::Get()->is_np_shape()) {
+    CHECK_EQ(storage_type(), kDefaultStorage)
+        << "only allow serializing ndarray of default storage type within the scope of np_shape";
+    CHECK_NE(shape_.Size(), 0U)
+        << "serializing zero-size ndarray within the scope of np_shape is not supported";
+    CHECK_NE(shape_.ndim(), 0)
+        << "serializing scalar ndarray within the scope of np_shape is not supported";
+  }
   // write magic number to mark this version
   // for storage type
   strm->Write(NDARRAY_V2_MAGIC);
```

Comment:
would there be any need to differentiate the two arrays?

Reply:
added a magic number to represent serializing the ndarray in np shape semantics.